torch/nn/parallel/_functions.py

```diff
@@ -1,11 +1,22 @@
 import warnings
 
 import torch
+import types
+import importlib
 from . import comm
 from torch.autograd import Function
-from torch._utils import _get_device_index
-from typing import List, Optional
-
+from torch._utils import _get_device_index, _get_available_device_type
+from typing import List, Optional, Tuple
+from functools import lru_cache
+
+@lru_cache(maxsize=None)
```

Comment:
This looks OK. Curious about the tradeoffs here. What are the pros and cons of this approach vs init `ddp_device_type` as a global var in `https://github.com/pytorch/pytorch/blob/master/torch/distributed/__init__.py`

Reply:
Comparing with global var, no one can override it with mistake?